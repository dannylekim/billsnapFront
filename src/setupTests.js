// setup file
/** 
 * DON'T TOUCH THIS FILE.
 * IT'S REQUIRED TO TEST COMPONENTS
*/
import Enzyme, { shallow, render, mount } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer  from 'react-test-renderer';

configure({ adapter: new Adapter() });

const matches = children => expect(
    renderer.create(children).toJSON()
  ).toMatchSnapshot();
  
global.shallow = shallow;
global.render = render;
global.mount = mount;
global.matches = matches;
global.localStorage = {
  setItem: jest.fn()
};