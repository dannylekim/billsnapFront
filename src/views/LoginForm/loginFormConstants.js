export const loginFormInputs = [
{
    placeholder: "Enter your email",
    name: "email",
    type: "email"
},
{
    placeholder: "Enter a password",
    name: "password",
    type: "password"
}
];
