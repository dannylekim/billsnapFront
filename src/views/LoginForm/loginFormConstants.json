[
    {
        "placeholder": "Enter your email",
        "name": "email",
        "type": "email",
        "autoComplete": "username"
    },
    {
        "placeholder": "Enter a password",
        "name": "password",
        "type": "password",
        "autoComplete": "new-password"
    }
]
