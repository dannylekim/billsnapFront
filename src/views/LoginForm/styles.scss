div.page__login {
    display: flex;
    justify-content: space-between;
}
div.register__account__container{
    width:70%;
    margin-left: auto;
    margin-right: auto;
    border: 1px solid black;
    background-color: white;
}

div.login__form .form-control{
    font-size:18px;
    // margin-top: 50px;
    width:70%;
    margin-left: auto;
    margin-right: auto;
    color: black;
    background-color: white;
}

a.forgot__password{
    margin-right: 10px;
    font-size: 12px;
    font-style: normal;
    font-weight: bold;
    color: #000000;
}

a.forgot__password:hover{
    text-decoration: none;
}

div.login__container{
    background-color: #FFFFFF;
    box-shadow: 0px 7.5px 80px rgba(90, 97, 105, 0.17);
    width: 500px;
    height: 480px;
    margin-left: auto;
    margin-right: auto;
  }