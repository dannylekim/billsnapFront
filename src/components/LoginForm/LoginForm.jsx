import React, { useState } from "react";
import PropType from "prop-types";
import {loginFormInputs} from "./loginFormConstants";
import {login} from "../../utils/requests/UserRequests";
import {Alert, Tooltip, Button, Form, FormGroup, FormInput} from "shards-react";
import "./styles.scss";

export const LoginForm = ({handleButtonClick ,onChange}) => {
    return (
        <div className= "login__form">
            <Form>
                <div className="form-inputs">
                    {loginFormInputs.map((inputs , key) => 
                    <FormGroup key={key} onChange={onChange}>
                            <FormInput  invalid={false} 
                                        valid={false} 
                                        className="mb-2" 
                                        type={inputs.type} 
                                        name={inputs.name} 
                                        id={inputs.name} 
                                        placeholder ={inputs.placeholder} />       
                        </FormGroup>  
                        )
                    }
                </div>
                <FormGroup>
                    <Button size="lg" pill theme="dark" onClick= {(event) => handleButtonClick(event)}  name="submit">{"Login"}</Button> 
                </FormGroup>
            </Form>
        </div>
    );
};

export default (props) => {
    const defaultErrors = {
        email:      {    
                        hasError:false,
                        message: "" 
                    },
        password:   {   hasError:false,
                        message: "" 
                    }   
    };

    const [user_credentials, setUserCredential] = useState({}); 
    const [error_message, setErrorMessage] = useState("");
    const [hasErrors, setHasErrors] = useState(defaultErrors);
    const [alertMessage, setAlertMessageFields] = useState({
        visible: false,
        alertType: "danger"
    });

    /**
     * Triggers the error alert banner.
     * !!!! Temporary put a success alert for successfull login. 
     * @param {String} alertType the alert type success or error.
     */
    const triggerAlert = (alertType) => {
        setAlertMessageFields({visible: !alertMessage.visible,
                               alertType: [alertType]});
    };
    /**
     * @function handleResponse
     * @description filters the error type and setting the erro message and form tip error.
     * @param {Object} response the Http response from backend.
     */
   const handleResponse = (response) => {
        if(response.status === "BAD_REQUEST"){
            let errors = response.errors;
            errors.map(error => (
                setHasErrors({...hasErrors, [error.field]: {...hasErrors[error.field], hasError: true , message: error.message}}) 
            ))
        };
  
        return setErrorMessage(response.message);
    };
    /**
     * @function handleButtonClick 
     * @description Handles the submission, calls the backend API, through login function.
     * handle the response with handleResponse function.
     * @param {Event} event 
     */
    const handleButtonClick = (event) => {
        event.preventDefault();
        login(user_credentials).then(response => {
            if(!response.token) { 
                triggerAlert("danger");
                handleResponse(response);
            }
            else {
                localStorage.setItem("token", response.token);
                triggerAlert("success");//temporary TODO: will be redirect to home. props.history.push('/home')
                setErrorMessage(response.message);//temporary
            };    
        });  
    };
    /**
     * @function onChange 
     * @description event handling function, handles the changes in the input fields.
     * @param {Event} event 
     */
    const onChange = (event) => {
        const {name, value} = event.target;
        setUserCredential((prev) => ({...prev, [name]: value}));
        setHasErrors(defaultErrors);
    };
    return (
    <div className="loginForm">
        <Alert dismissible={triggerAlert} open={alertMessage.visible} className="mb-3" theme={alertMessage.alertType}>
            {error_message}
        </Alert>
      <LoginForm handleButtonClick={handleButtonClick} onChange={onChange}/>
        {/* TODO: onChange error handling */}
        {["email", "password"].map((field, key) => (
            <Tooltip
                key={key}
                open={hasErrors[field].hasError}
                target={`#${field}`}
                >
               <span id="input_error"> {hasErrors[field].message} </span>
            </Tooltip>
        ))
        }
    </div>
    );
};

LoginForm.propTypes = {
    handleButtonClick: PropType.func.isRequired
  };