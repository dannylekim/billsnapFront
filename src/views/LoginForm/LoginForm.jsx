import React, {Component} from "react";
import PropType from "prop-types";
import loginFormInputs from "./loginFormConstants.json";
import {login} from "../../utils/requests/UserRequests";
import {Alert, Button, Form, FormGroup, FormInput, Tooltip,} from "shards-react";
import Loader from "../../components/Loader";

import "./styles.scss";

export const LoginForm = ({
  handleButtonClick,
  onChange,
  hasErrors,
  alertMessage,
  dismissAlert,
  error_message,
  setFormType,
}) => {
  return (
    <div>
      {alertMessage.visible === true ? (
        <Alert
          dismissible={dismissAlert}
          open={alertMessage.visible}
          className="mb-3"
          theme={alertMessage.alertType}
        >
          {error_message}
        </Alert>
      ) : (
        <div className="hidden__div" />
      )}

      <img
        alt="character logo"
        src="./billSnapIcon.png"
        className="character__icon__image"
      />

      <Form>
        <div className="form__inputs">
          {loginFormInputs.map((inputs, key) => (
            <FormGroup key={key} onChange={onChange}>
              <FormInput
                className="register__login__inputs"
                type={inputs.type}
                name={inputs.name}
                id={inputs.name}
                placeholder={inputs.placeholder}
                autoComplete={inputs.autoComplete}
                invalid={hasErrors[inputs.name].hasError}
              />
            </FormGroup>
          ))}
        </div>
        <div className="forgot__password__login">
          <a href="/#" className="forgot__password">
            Forgot Password?
          </a>

          <Button
            size="md"
            pill
            className="login_register__submit__button"
            onClick={handleButtonClick}
            name="submit"
          >
            Log in
          </Button>
        </div>
      </Form>

      {loginFormInputs.map((field, key) => (
        <Tooltip
          key={key}
          placement="left"
          open={hasErrors[field.name].hasError}
          target={`#${field.name}`}
        >
          <span id="input_error"> {hasErrors[field.name].message} </span>
        </Tooltip>
      ))}

      <div>
        <div className="form__seperator">
          <hr className="form__horizontal__line" />
          Or
          <hr className="form__horizontal__line" />
        </div>
        <div>
          <h6>New to Billsnap?</h6>
          <Button className="form__toggle" onClick={setFormType}>
            {"ðŸ‘‹ Create account"}
          </Button>
        </div>
      </div>
    </div>
  );
};

export const DEFAULT_ERRORS = {
  email: { hasError: false, message: "" },
  password: { hasError: false, message: "" },
};

export const DEFAULT_ALERT_MESSAGE = {
  visible: false,
  alertType: "danger",
};

class LoginFormContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      user_credentials: {
        email: "",
        password: "",
      },
      error_message: "",
      hasErrors: DEFAULT_ERRORS,
      alertMessage: DEFAULT_ALERT_MESSAGE,
      isLoading: false,
    };

    this.dismissAlert = this.dismissAlert.bind(this);
    this.handleErrorResponse = this.handleErrorResponse.bind(this);
    this.handleSubmitClick = this.handleSubmitClick.bind(this);
    this.onFormChange = this.onFormChange.bind(this);
  }

  /**
   * @function dismissAlert
   * @description dismisses the alert message. and closes tool tip.
   */
  dismissAlert = () => {
    this.setState((prev) => ({
      alertMessage: {
        ...prev.alertMessage,
        visible: false,
      },
      hasErrors: DEFAULT_ERRORS,
    }));
  };

  /**
   * @function handleResponse
   * @description filters the error type and setting the erro message and form tip error.
   * @param {Object} response the Http response from backend.
   */
  handleErrorResponse = (response) => {
    if (response.status === "BAD_REQUEST") {
      response.errors.forEach((error) =>
        this.setState((prev) => ({
          hasErrors: {
            ...prev.hasErrors,
            [error.field]: {
              hasError: true,
              message: error.message,
            },
          },
        }))
      );
    }

    if (response.status === "UNAUTHORIZED") {
      this.setState({
        hasErrors: {
          email: {
            hasError: true,
            message: "email might not exist in system.",
          },
          password: { hasError: true, message: "forgot your password?" },
        },
      });
    }

    if (
      response.status === "BAD_REQUEST" ||
      response.status === "UNAUTHORIZED"
    ) {
      this.setState({
        alertMessage: { visible: true, alertType: "danger" },
      });
    }

    return this.setState({
      error_message: response.message,
    });
  };

  /**
   * @function handleSubmitClick
   * @description Handles the submission, calls the backend API, through login function.
   * handle the response with handleResponse function.
   * @param {Event} event
   */
  handleSubmitClick = async (event) => {
    event.preventDefault();
    this.dismissAlert();
    try {
      this.setState({
        isLoading: true,
      });
      const response = await login(this.state.user_credentials);
      if (!response.token) {
        this.handleErrorResponse(response);
      } else {
        localStorage.setItem("billSnap_token", response.token);
        const { id, ...userProfile } = response.profile;
        this.props.setUser(userProfile);
        this.props.history.push("/dashboard");
      }
    } catch (error) {
      this.handleErrorResponse(error);
    } finally {
      this.setState({
        isLoading: false,
      });
    }
  };

  /**
   * @function onFormChange
   * @description event handling function, handles the changes in the input fields.
   * @param {Event} event
   */
  onFormChange = (event) => {
    const { name, value } = event.target;
    this.setState((prev) => ({
      user_credentials: {
        ...prev.user_credentials,
        [name]: value,
      },
      hasError: DEFAULT_ERRORS,
    }));
  };

  render() {
    const {
      hasErrors,
      user_credentials,
      alertMessage,
      error_message,
    } = this.state;

    return (
      <div className="login__container">
        {this.state.isLoading && <Loader />}
        <LoginForm
          handleButtonClick={this.handleSubmitClick}
          onChange={this.onFormChange}
          hasErrors={hasErrors}
          user_credentials={user_credentials}
          alertMessage={alertMessage}
          dismissAlert={this.dismissAlert}
          error_message={error_message}
          setFormType={this.props.setFormType("register")}
        />
      </div>
    );
  }
}

LoginFormContainer.propTypes = {
  setFormType: PropType.func.isRequired,
};

export default LoginFormContainer;
