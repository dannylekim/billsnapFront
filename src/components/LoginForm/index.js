import LoginForm from "./LoginForm";
export * from "./LoginForm.jsx";
export default LoginForm;