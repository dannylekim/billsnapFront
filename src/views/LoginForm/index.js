import LoginForm from "./LoginForm";
import {connect} from "react-redux";
import {setUser} from "../../redux/actions/userActions";

const mapDispatchToProps = (dispatch) => ({
  setUser: (user = {}) => dispatch(setUser(user)),
});

export default connect(null, mapDispatchToProps)(LoginForm);
