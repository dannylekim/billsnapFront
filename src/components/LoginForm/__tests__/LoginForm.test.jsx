import React from 'react';
import { LoginForm } from '../LoginForm.jsx';
import { shallow } from "enzyme";
import {matches} from "../../../setupTests";

describe('LoginForm', () => {
    let wrapper;
    let handleMockFunction;
  
    beforeEach(() => {
      handleMockFunction = jest.fn();
      wrapper = shallow(
        <LoginForm handleButtonClick={handleMockFunction} />
      );
    });
  
    afterEach(() => {
      handleMockFunction.mockRestore();
    });
  
    describe("render", () => {
      describe("snapshots ðŸ“¸", () => {
        it("Login should match snap shot", () => {
          matches(<LoginForm handleButtonClick={handleMockFunction}/>);
        });
      });
  
      describe("components", () => {});
    });
  
    describe("functions", () => {});
});