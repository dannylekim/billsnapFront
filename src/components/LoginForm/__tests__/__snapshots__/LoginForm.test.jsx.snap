// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm render snapshots ðŸ“¸ Login should match snap shot 1`] = `
<div
  className="login__form"
>
  <form
    className=""
  >
    <div
      className="form-inputs"
    >
      <div
        className="form-group"
      >
        <input
          className="mb-2 form-control"
          id="email"
          name="email"
          placeholder="Enter your email"
          type="email"
        />
      </div>
      <div
        className="form-group"
      >
        <input
          className="mb-2 form-control"
          id="password"
          name="password"
          placeholder="Enter a password"
          type="password"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <button
        className="btn btn-dark btn-lg btn-pill"
        name="submit"
        onClick={[Function]}
        type="button"
      >
        Login
      </button>
    </div>
  </form>
</div>
`;
